### PR TITLE
Update vellum's datasources structure

### PR DESCRIPTION
**Pull request by orangejenny**
screen shot 2016-06-29 at 5 46 09 pm

### CODE REVIEW

File: corehq/apps/app_manager/views/formdesigner.py [hunk 1]
```diff
@@ -199,7 +199,7 @@ def get_form_data_schema(request, domain, form_unique_id):
     try:
         data.append(get_session_schema(form))
         if form and form.requires_case():
-            data.append(get_casedb_schema(app))  # TODO use domain instead of app
+            data.append(get_casedb_schema(app, form.get_module().case_type))  # TODO use domain instead of app
```
Comment (emord) [inline corehq/apps/app_manager/views/formdesigner.py]: Unrelated, but I don't think this comment should be here. app seems to be the right context here

File: corehq/apps/app_manager/tests/test_util.py [hunk 2]
```diff
@@ -119,10 +121,21 @@ def test_get_casedb_schema_with_related_case_types(self):
...
+        self.add_form("referral")
+        child = self.add_form("child")
+        self.factory.form_opens_case(child, case_type='referral', is_subcase=True)
+        pregnancy = self.add_form("pregnancy")
+        self.factory.form_opens_case(pregnancy, case_type='referral', is_subcase=True)
+        schema = util.get_casedb_schema(self.factory.app, 'referral')
         subsets = {s["id"]: s for s in schema["subsets"]}
-        self.assertEqual(subsets["village"]["related"], None)
-        self.assertDictEqual(subsets["family"]["related"], {"parent": "village"})
+        self.assertTrue(re.match(r'^parent \((pregnancy|child) or (pregnancy|child)\)$', subsets["parent"]["name"]))
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_util.py]: E501 line too long (116 > 115 characters)

File: corehq/apps/app_manager/tests/test_util.py [hunk 3]
```diff
@@ -119,10 +121,21 @@ def test_get_casedb_schema_with_related_case_types(self):
...
+        self.assertEqual(subsets["parent"]["structure"], {"case_name":{}})
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_util.py]: E231 missing whitespace after ':'

**Comment by orangejenny** [comment]
@emord / @millerdev 

Ready for review. Produces data source like
```
{
    "name": "case",
    "subsets": [
      {
        "structure": {
          "name": { },
          "case_name": { }
        },
        "related": {
          "parent": "parent"
        },
        "id": "case",
        "key": "@case_type",
        "name": "child_type"
      },
      {
        "structure": {
          "parent_thing": { },
          "stepparent_thing": { },
          "case_name": { }
        },
        "related": {
          "parent": "grandparent"
        },
        "id": "parent",
        "key": "@case_type",
        "name": "parent (stepparent_type or parent_type)"
      },
      {
        "structure": {
          "grandparent_thing": { },
          "case_name": { }
        },
        "related": null,
        "id": "grandparent",
        "key": "@case_type",
        "name": "grandparent (grandparent_type)"
      }
    ],
    "uri": "jr:\/\/instance\/casedb",
    "path": "\/casedb\/case",
    "id": "casedb",
    "structure": { }
  }
```

File: corehq/apps/app_manager/util.py [hunk 4]
```diff
@@ -318,28 +321,52 @@ def get_all_case_properties(app):
...
+        def _add_ancestors(ctype, generation):
+            if generation < len(generation_names):
+                generations[generation].append(ctype)
+                for parent in related.get(ctype, {}).get('parent', []):
+                    _add_ancestors(parent, generation + 1)
+
+        _add_ancestors(base_case_type, 0)
+
+        # Remove any duplicate types or empty generations
+        generations = [set(g) for g in generations if len(g)]
```
Comment (emord) [inline corehq/apps/app_manager/util.py]: I think you can get rid of this if you use `generations = [set() for g in generation_names]` above

Comment (orangejenny) [inline corehq/apps/app_manager/util.py]: that takes card of the `set(g)` part but not the `if len(g)`

File: corehq/apps/app_manager/tests/test_util.py [hunk 5]
```diff
@@ -93,36 +94,56 @@ class SchemaTest(SimpleTestCase):
     def setUp(self):
         self.factory = AppFactory()
 
-    def test_get_casedb_schema_empty_app(self):
```
Comment (emord) [inline corehq/apps/app_manager/tests/test_util.py]: Should this test be replaced with a test for a form that doesn't use case management?

File: corehq/apps/app_manager/util.py (hunk 4 above)
Comment (emord) [inline corehq/apps/app_manager/util.py]: Ah. missed that. I suppose you could move it down to the return statement, but that's already a nasty nested list/dictionary comprehension

**Comment by emord** [comment]
Nothing blocking for me.

 @millerdev any thoughts on our proposed changes to the schema? Happy to offline if it's not obvious from this PR to https://github.com/dimagi/Vellum/pull/684

File: corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    def _form_error(self, form, error=""):
+        logger.error("{} (domain {}, app {}, form {})".format(error, form.get_app().domain,
+                                                              form.get_app().id, form.unique_id))
+
+    def _replace_in_form(self, form, relationships, case_type, affix_index):
+        if not case_type:
+            return
+        if affix_index < len(self.affixes):
+            #logger.info("Replacing #case/{}/ with #case/{}".format(case_type, self.affixes[affix_index]))
+            form.source = form.source.replace("#case/{}/".format(case_type), "#case/{}".format(self.affixes[affix_index]))
```
Comment (dimagimon) [inline corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py]: E501 line too long (122 > 115 characters)

File: corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import logging
+import re
+
+from django.core.management import call_command
+from django.core.management.base import BaseCommand
+
+from corehq.toggles import all_toggles
+from corehq.apps.domain.models import Domain
+from corehq.apps.app_manager.dbaccessors import get_app, get_apps_in_domain
```
Comment (dimagimon) [inline corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py]: F401 'get_app' imported but unused

File: corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
+import logging
+import re
+
+from django.core.management import call_command
```
Comment (dimagimon) [inline corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py]: F401 'call_command' imported but unused

**Comment by orangejenny** [comment]
@emord added migration script

File: corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            elif len(parents) == 1:
+                self._replace_in_form(form, relationships, parents[0], affix_index + 1)
+        else:
+            self._form_error(form, "Hierarchy too deep")
+
+    def handle(self, *args, **options):
+        toggle_map = dict([(t.slug, t) for t in all_toggles()])
+        domains = [row['key'] for row in Domain.get_all(include_docs=False)]
+        for domain in domains:
+            if toggle_map['rich_text'].enabled(domain) or toggle_map['experimental_ui'].enabled(domain):
```
Comment (emord) [inline corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py]: I think you also need to see if domain has `vellum_case_management` enabled

File: corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                        for form in module.forms:
+                            if form.doc_type != 'AdvancedForm' and form.requires_case():
+                                #logger.info('migrating form {}'.format(form.name.get('en', form.name)))
+                                base_case_type = form.get_module().case_type
+                                self._replace_in_form(form, relationships, base_case_type, 0)
+                                prefixes = re.findall(r'#case/\w+/', form.source)
+                                prefixes = set(prefixes)
+                                for p in prefixes:
+                                    if p != "#case/parent/" and p != "#case/grandparent/":
+                                        self._form_error(form, "Unknown prefix remaining: {}".format(p))
```
Comment (emord) [inline corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py]: What do you think about adding a `dry-run` option to this that won't save?

**Comment by millerdev** [comment]
> any thoughts on our proposed changes to the schema? Happy to offline if it's not obvious from this PR to dimagi/Vellum#684

I like that the case type is no longer in the hashtag path. As I mentioned yesterday, I'm not too fond of `grandparent`, but I suppose we can deprecate it in favor of a better syntax if it becomes a problem in the future.

Will it work if a case has a property named `parent` or `grandparent`? If the case had a property like that then the `parent` part of these two paths have very different meanings:
```
Reference a "parent" property on the session case:
#case/parent

Reference a "dob" property on the session case's parent case:
#case/parent/dob
```
Will that matter? I think it will not matter at all in the schema—it should be able to handle a situation like that. However, I'm less sure about the vellum code.

**Comment by emord** [comment]
@millerdev works in the vellum code since

image

File: corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py (hunk 9 above)
Comment (orangejenny) [inline corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py]: https://github.com/dimagi/commcare-hq/pull/11794 hasn't been run yet. Are there apps that you've manually migrated that are in domains that don't have the flags turned on?

File: corehq/apps/app_manager/tests/test_util.py (hunk 5 above)
Comment (orangejenny) [inline corehq/apps/app_manager/tests/test_util.py]: yes
https://github.com/dimagi/commcare-hq/pull/12252/commits/dbd858733feb4e75f4dc1fba939aa6d9338b255d

File: corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py (hunk 10 above)
Comment (orangejenny) [inline corehq/apps/app_manager/management/commands/cmitfb_migrate_syntax.py]: good though => https://github.com/dimagi/commcare-hq/pull/12252/commits/a1feb40dbcd13c0408775efcf381aceb67842787

**Comment by millerdev** [comment]
👍

**Comment by emord** [comment]
looks related 

```
======================================================================
FAIL: test_get_session_schema_advanced_form (corehq.apps.app_manager.tests.test_util.SchemaTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock.py", line 1224, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_util.py", line 215, in test_get_session_schema_advanced_form
    "key": "@case_id",
AssertionError: {'reference': {'subset': 'case', 'source': 'casedb', 'key': '@case_id'}} != {'reference': {'subset': 'visit', 'source': 'casedb', 'key': '@case_id'}}
- {'reference': {'key': '@case_id', 'source': 'casedb', 'subset': 'case'}}
?                                                                  ^^ ^
+ {'reference': {'key': '@case_id', 'source': 'casedb', 'subset': 'visit'}}
?                                                                  ^^ ^^
----------------------------------------------------------------------
```

**Comment by millerdev** [comment]
retracting my 👍 see comment on the [vellum pr](https://github.com/dimagi/Vellum/pull/684/files/bf09e0abfcc2e7799567de63bca7f4ed1fd36a40#r69141452)

**Comment by orangejenny** [comment]
Most recent commit should fix tests, but unfortunately also removes most of the tests added by https://github.com/dimagi/commcare-hq/pull/11877 (fyi @snopoke ) ...because the new format uses relationship instead of case type as the id for each subset, and the supported relationships don't cover everything that can be done in advanced modules.

**Comment by millerdev** [comment]
Why did those tests need to be removed? Does `util.get_session_schema(...)` no longer return something that makes sense for them?

**Comment by orangejenny** [comment]
@millerdev Yeah - because the subset ids are now `case`, `parent`, etc. instead of case type, I'm lazily having `get_session_schema` always use `case` as the subset id (https://github.com/dimagi/commcare-hq/pull/12252/commits/2027dc075bcf71abce19ee5afeb706d95a038374#diff-8a78f7edae07ff768accdb85fbaace47). The new structure just flat out doesn't support the kind of case hierarchy that was being tested in `test_get_session_schema_advanced_form_multiple_cases` (https://github.com/dimagi/commcare-hq/pull/12252/files#diff-b9de4a695c074e8e8121b6e74f402a9dL186).

Happy to offline - I'm not crazy about making this change, as deleting tests feels like a step backwards, but on the other hand we're not officially supporting advanced forms at all at this point.

**Comment by millerdev** [comment]
> The new structure just flat out doesn't support the kind of case hierarchy that was being tested in `test_get_session_schema_advanced_form_multiple_cases`

Seems like we're heading in the wrong direction if that's true. The schema should be able to represent any kind of relationship diagram that can be created in HQ. It's fine if other front-end tools like Vellum's case management tree view choose not to represent it accurately or completely, but not fine if we deliberately make our schema incapable of representing valid structures that we know can exist.

**Comment by millerdev** [comment]
We offlined, and got me up to speed on a few things that I'd overlooked or wasn't understanding completely. I'm back in the game with a 👍

**Pull request merged**